src/main.rs

```diff
@@ -336,6 +399,34 @@ async fn forward_request(
 
         builder.body(body)?
     } else {
+        let body_valid = match (certificate, tree) {
+            (Some(Ok(certificate)), Some(Ok(tree))) => match validate_body(
+                &certificate,
+                &tree,
+                &canister_id,
+                &agent,
+                &uri,
+                &http_response.body,
+                logger.clone(),
+            ) {
+                Ok(valid) => valid,
+                Err(e) => {
+                    return Ok(Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(format!("Certificate validation failed: {}", e).into())
+                        .unwrap());
+                }
+            },
+            (Some(_), _) | (_, Some(_)) => false,
+            // Canisters don't have to provide certified variables
+            (None, None) => true,
```

Comment:
this is problematic as @nomeata has pointed out. The service worker code 
https://github.com/dfinity/ic/blob/master/typescript/service-worker/src/sw/http_request.ts#L287
treats the absence of IC-CERTIFICATE as invalid responses. Here the icx-proxy is treating it as the valid body. So there is a  semantic difference in this case.

// Canisters don't have to provide certified variables
Given this comment, I think we need a bit more discussion around this.

Reply:
The long term plan is to close this gap, but in order to not break existing canisters, this fallback has to remain in place for now.